### ISSUE TITLE
Spark 3.* Cosmos DB connector schema inference issue for nested arrays

### ISSUE THREAD

**Issue by hmm100** [Azure/azure-sdk-for-java#26503]
For the past few days I've been trying to use the _azure-cosmos-spark_3-1_2-12_ connector (version 4.5.3), using different schema inference configuration options but it's not working as expected.

Tested on Databricks Runtime Version 8.4 and 9.1 LTS.

**Full Collection**
Code used to bring over the entire collection (~820 documents):
``` python
readConfig = {
  "spark.cosmos.accountEndpoint": CosmosDBEndpoint,
  "spark.cosmos.accountKey": CosmosDBMasterKey,
  "spark.cosmos.database": CosmosDBDatabase,
  "spark.cosmos.container": CosmosDBCollection,
  "spark.cosmos.read.inferSchema.includeSystemProperties": "true"
}

# Connect via azure-cosmosdb-spark to create Spark DataFrame
dfCosmosDB = (spark.read
		   .format("cosmos.oltp")
		   .options(**readConfig)
		   .load())

dfCosmosDB.printSchema()
```

The following is part of the resulting schema that matters to this discussion. For ~100 of those documents, the `Actions.Condition` is filled in with data, but the connector infers it as `NullType` and forces every result to null:
![FullCollection_Schema](https://user-images.githubusercontent.com/97683370/149538332-3d7cd286-38ff-4322-a184-0eea9a920778.png)

**Infer Schema Query**
Used the `inferSchema.query` option to force the connector to only use documents that have the `Actions.Condition` filled in (including filtering by a specific id), but it didn't change the result:
``` python
"spark.cosmos.read.inferSchema.query": f"SELECT f FROM f JOIN a IN f.Actions JOIN c IN a.Condition"
```
![InferSchemaQuery_Schema](https://user-images.githubusercontent.com/97683370/149540005-7780b5a9-4027-42c5-9d79-88e544545ff5.png)

**Custom Query**
Tried as well to use the `customQuery` to fetch only documents that have the `Actions.Condition` populated, but no change in the resulting schema:
``` python
"spark.cosmos.read.customQuery": f"SELECT f FROM f JOIN a IN f.Actions JOIN c IN a.Condition"
```
![CustomQuery_f_Schema](https://user-images.githubusercontent.com/97683370/149540849-f296e09c-de54-4198-8015-6613a498a5ea.png)

If you look at the query used above, it's fetching data from the root `f`, but if it's changed to one of the nested arrays, e.g. `a`, the schema will be inferred correctly:
``` python
"spark.cosmos.read.customQuery": f"SELECT a FROM f JOIN a IN f.Actions JOIN c IN a.Condition"
```
![CustomQuery_a_Schema](https://user-images.githubusercontent.com/97683370/149541918-d726ea16-a30c-4d30-bef2-2d59b1e19795.png)

This change in behavior occurs both on `customQuery` and `inferSchema.query` options.
Important to note that for all documents that have `Actions.Condition` populated, also have empty/null `Actions.Filter`, which is why it's showing as `NullType` (correctly) in the last schema. This also makes me wonder why in the previous query it was showing a structure, as if it was using documents outside of the scope of the query to infer the schema.

**Comment by FabianMeiswinkel** [comment]
Are you able to share an anonimized sample document here (or via mail to fabianm@microsoft.com)?

Thanks,
Fabian

**Comment by hmm100** [comment]
> Are you able to share an anonimized sample document here (or via mail to [fabianm@microsoft.com](mailto:fabianm@microsoft.com))?
> 
> Thanks, Fabian

Hi @FabianMeiswinkel,
As I only consume the data, I'll have to ask the team responsible for a sample document, but at this time everyone is already gone for the weekend.
Monday I'll try to provide the requested sample.

Kind Regards,
Hugo

**Comment by ealsur** [comment]
Following the logical path:

1. Assuming the inference starts on the `Actions` property, which we assume it's an `ArrayNode`: https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferrer.scala#L194-L197
2. It peeks into the Array and infers the item type checking the first element in the array: https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferrer.scala#L213 - **this is one key point, if the array has elements with different schema, it will only look at the first**
3. For that first item, it again attempts to infer the type, assuming that it executes https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferrer.scala#L176-L210 for the `Condition` property.
4. At this point, the only way `Condition` is null is if:
  1. The value is `null` for that sampled document.
  2. Or it's an Array where the first element is `null` - This is my hunch, which we can clarify with sample documents.

**Comment by hmm100** [comment]
Please find below a sample document as requested.

[SampleDocument.txt](https://github.com/Azure/azure-sdk-for-java/files/7881269/SampleDocument.txt)

Looking at @ealsur reply, I think the point 6 (Array where the first element is `null`) is spot on.
This document basically holds the configuration of a flow, so the first element of Actions rarely (if ever) will have Conditions associated.
If only one element is looked at for arrays, another problem rises, which is the Filter and Condition which are rarely populated together. As of now, no documents include Filter and Condition for the same Action so it would be impossible for the infer schema to correctly detect both of them. One of them would always come up empty.

**Comment by ealsur** [comment]
Based on your example, the other problem (besides the first element being `null`) is that in the document the property is an empty array:

```
"Actions": [
        {
            "Type": "http-request",
            "Version": 1,
            "Label": null,
            "Filter": [],
            "Condition": [],
```

and then in another item inside `Actions`:


```
{
            "Type": "http-request",
            "Version": 1,
            "Label": null,
            "Filter": [],
            "Condition": [
                {
                    "BodyPath": "{{ step1.StatusCode }}",
                    "Operation": 2,
                    "Value": "400"
                }
            ],
```

So the same array property can be empty in some subitem and not empty on others. The problem with empty arrays is that we cannot infer the type (what type are they elements?). 

Potentially we can attempt to find if there is anything that is not empty and use that.

**Comment by ealsur** [comment]
@hmm100 based on the conversation and the example, I was able to find the cases where the inference could be `NullType` on your case, see the linked PR.

**Comment by hmm100** [comment]
> @hmm100 based on the conversation and the example, I was able to find the cases where the inference could be `NullType` on your case, see the linked PR.

If I understood the PR correctly, after it has been reviewed and approved, the inference process will start looking into all elements of an array instead of just the first. Is that correct?

**Comment by ealsur** [comment]
Correct, I even used part of your sample data into the tests and the PR description that explains how it would work.
